Add most in-place references/decompositions

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88117
* #88116
* #88115

We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

lezcano committed:
Add most in-place references/decompositions

We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

ezyang commented on test/test_meta.py:
```diff
@@ -961,6 +961,13 @@ def __torch_function__(self, func, types, args=(), kwargs=None):
+
+meta_inplace_skips = set({
+    # Errors out in one of the tests, while ProxyTensor passes...
+    torch.Tensor.cumsum_,
```
What's going on here?

lezcano replied:
I don't know. I had to fix so many errors in this PR that at some point I was too tired to keep digging. I'll investigate it and add a PR fixing that. I think the issue may be that proxy tensor may not be run on integral types, while this test is. Then, if you run the in-place variant on integral types with a `dtype=` that's a float, it breaks. It may be an issue of coverage.

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -424,6 +425,30 @@ def _fn(*args, **kwargs):
 
+def _make_inplace(fn):
+    """
+    Given a function with out variant (i.e. using `out_wrapper()), it returns its in-place variant
```
Should we point the reader to the wiki faq discussing the semantics of this?

lezcano replied:
Sure thing

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -424,6 +425,30 @@ def _fn(*args, **kwargs):
+    Given a function with out variant (i.e. using `out_wrapper()), it returns its in-place variant
+    """
+
+    # nb. We use the name of the first argument used in the unary references
```
Style nit: boo to adding more Italian

lezcano replied:
I literally learnt of this expression in the last years, while reading through the PyTorch codebase. You'll have to point your finger to other people if you want to find the root of the issue :P

mruberry commented on torch/_refs/__init__.py:
```diff
@@ -4945,6 +4970,88 @@ def bucketize(
+trunc_ = _make_inplace(trunc)
+xlogy_ = _make_inplace(xlogy)
+
+# views
```
This should probably add a comment for why inplace view operations can't be modeled using the current technique

mruberry commented on torch/_refs/nn/functional/__init__.py:
```diff
@@ -123,39 +150,46 @@ def dropout(
-        rhs = torch.expm1(a)
-
-    return torch.where(a > 0, a, rhs)
+    # nb. This should be factored out into a can_cast aux function
```
Nice revision

mruberry reviewed (approved):
Looks good; see inline comments.

We discussed inplace view operations offline, and I think leaving them for a future PR makes sense. We should define their semantics and update the Wiki before implementing something for them.

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano committed:
Update on "Add most in-place references/decompositions"


We add most in-place references in a generic way. We also implement a
wrapper to implement the annoying interface that `nn.functional`
nonlinearities have.

We fix along the way a couple decompositions for some non-linearities by
extending the arguments that the references have.

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano commented:
@pytorchbot merge -f "At least one of the two failures is unrelated. Let's hope that the Torchbench one is also unrelated"
